### PR TITLE
AJ `perform_enqueued_jobs` shouldn't perform job retries: 

### PR DESCRIPTION
**Pull request by Edouard-chin**
AJ `perform_enqueued_jobs` shouldn't perform job retries:

- ### Problem

  If we use `perform_enqueued_jobs` without a block, a job that
  uses a retry mechanism to reeenqueue itself would get performed
  right away.
  This behaviour make sense when using `perform_enqueued_jobs` with
  a block.

  However I'm expecting `perform_enqueued_jobs` without a block to
  perform jobs that are **already** in the queue not the ones that
  will get enqueued afterwards.

  ### Solution

  Dup the array of jobs given to avoid future mutation.

cc/ @etiennebarrie @rafaelfranca @casperisfine

### CODE REVIEW

**Comment by rafaelfranca** [comment]
I don't think it is good to have different behavior based on if you pass a block or not. An inconsistent behavior is harder to understand if the block and non-block version behaves in the same way it is easier to understand.

**Comment by Edouard-chin** [comment]
But isn't that more confusing that `AJ::TestAdapter.perform_enqueued_jobs` being set to false actually perform jobs ? 

The fact that we iterate over an array that get mutated as we go makes it hard to understand why a job is getting performed where it shouldn't. Also the documentation states that if no block is passed, enqueued jobs will get performed up to this point, which makes me think that the behaviour is already different anyway https://github.com/rails/rails/blob/2a0823ecbaeb3a6fbedc5cc31879d68fdf27d0cb/activejob/lib/active_job/test_helper.rb#L496-L498

File: activejob/test/cases/test_helper_test.rb [hunk 1]
```diff
@@ -909,6 +909,27 @@ def test_perform_enqueued_jobs_block_with_at_with_job_wait_in_future
...
   end
 
+  def test_perform_enqueued_jobs_properly_count_job_that_raises
+    RaisingJob.perform_later("NotImplementedError")
+
+    assert_raises(NotImplementedError) do
+      perform_enqueued_jobs(only: RaisingJob)
+    end
+
+    assert_equal(1, performed_jobs.size)
```
[the remark this answers is not in the document]
Comment (Edouard-chin) [reply activejob/test/cases/test_helper_test.rb]: I can remove it , it's not a strict necessity. But FYI, when using the block version of `perform_enqueued_jobs`, performed_jobs would count all jobs even failing one https://github.com/rails/rails/blob/2f1fefe456932a6d7d2b155d27b5315c33f3daa1/activejob/lib/active_job/queue_adapters/test_adapter.rb#L53-L54

**Comment by jeremy** [comment]
This matches the stated intent of #33626, too.

It is a significant and subtle change, however. The method originally meant something like "with performance of these jobs enabled during the block, do …". But the non-block form followed the method name's literal meaning: "perform the specific jobs that have been enqueued before now." And it had a bug, arguably, in that it performed subsequent jobs as well, not just existing jobs.

That suggests these could be separate methods, distinguishing actions from behaviors. "Perform just the pending enqueued jobs" from "run matching jobs during this block."

Do tests for job retries, subsequent job enqueues like job hierarchies, and scheduled jobs need to wrap in a second block-form invocation now?
```ruby
# Do stuff that enqueues A jobs. A jobs enqueue B jobs as well.
A.perform_later

# OK, time to perform the work and assert expected outcomes
perform_enqueued_jobs only: [ A, B ]

# Oops, that didn't run the B job that A enqueued.
assert some_b_thing # => failed

# Do the work enqueued, and do the work that the work enqueues while doing the work?
perform_enqueued_jobs do
  perform_enqueued_jobs
end
```

**Comment by Edouard-chin** [comment]
> That suggests these could be separate methods, distinguishing actions from behaviors

Agreed, it will make things clearer if we had 2 distinct methods.

```ruby
# Do the work enqueued, and do the work that the work enqueues while doing the work?
perform_enqueued_jobs do
  perform_enqueued_jobs
end
```

This syntax might sound strange, though, to me, it make sense. However as you suggested, it would be much clearer if both methods didn't have the same name.

That being said, for clarity purposes, I'd say that if you have job A enqueuing Job B and have an assertion resulting on the outcome of B, then `perform_enqueued_jobs` without a block isn't suited for this case. Wrapping the `A.perform_later` inside `perform_enqueued_jobs(only: [A, B]) { ... }` is less verbose

**Comment by kaspth** [comment]
This now also seems to be in competition with https://github.com/rails/rails/pull/37685, and I'd prefer the deprecation to ultimately simplify rather than add this subtle change.

**Comment by doits** [comment]
Here's a simple helper method that runs all jobs, no matter if enqueued in background or not – maybe it helps someone who needs the old behavior:

```rb
def perform_all_enqueued_jobs
  perform_enqueued_jobs while enqueued_jobs.present?
end
```

**Pull request merged**